File: src/ImageCompression/Utils/Objects/OpcConvertor.java
package ImageCompression.Utils.Objects;

import ImageCompression.Containers.DataOpc;
import ImageCompression.Containers.Flag;
import ImageCompression.Containers.Size;
import ImageCompression.Utils.Functions.DCTMultiThread;
import ImageCompression.Utils.Functions.OPCMultiThread;

import static ImageCompression.Utils.Functions.OPCMultiThread.SIZEOFBLOCK;

public class OpcConvertor {
    enum State{Opc,Origin}
    final State state;
    short[][] dataOrigin;
    DataOpc[][] dataOpcs;
    Flag flag;
    boolean isReady=false;

    public OpcConvertor(short[][] dataOrigin, Flag flag) {
        this.dataOrigin = dataOrigin;
        this.flag = flag;
        state=State.Origin;
    }

    public OpcConvertor(DataOpc[][] dataOpcs, Flag flag) {
        this.dataOpcs = dataOpcs;
        this.flag = flag;
        state=State.Opc;
    }

    private DataOpc[][] directOPC(short[][]dataOrigin){

        Size size=sizeOpcCalculate(dataOrigin.length, dataOrigin[0].length);
        int duWidth= size.getWidth();
        int duHeight= size.getHeight();
        int Width=dataOrigin.length;
        int Height=dataOrigin[0].length;

        DataOpc[][]dopc=new DataOpc[duWidth][duHeight];
        short[][]buf=new short[SIZEOFBLOCK][SIZEOFBLOCK];
        for (int i = 0; i < duWidth; i++) {
//            System.out.print(Thread.currentThread().getId());
            for (int j = 0; j < duHeight; j++) {

                for (int x = 0; x < SIZEOFBLOCK; x++) {
                    for (int y = 0; y < SIZEOFBLOCK; y++) {
                        short value = 0;
                        int curX=i * SIZEOFBLOCK + x;
                        int curY=j * SIZEOFBLOCK + y;
                        if (curX< Width && curY < Height)
                            value = dataOrigin[curX][curY];

//                        if(x!=0||y!=0)
//                            assert value<0xff:"value["+curX+"]["+curY+"]="+value;
                        buf[x][y] = value;
                        // DU[i][j].setValue(val,x,y);
                    }
                }
                dopc[i][j]= OPCMultiThread.getDataOpc(buf, flag);
            }
        }
        return dopc;
    }
    private short[][] reverceOPC(DataOpc[][]dopc){

        int duWidth=dopc.length;
        int duHeight=dopc[0].length;
        int Width=duWidth*SIZEOFBLOCK;
        int Height=duHeight*SIZEOFBLOCK;
        short[][] res=new short[Width][Height];

//        boolean DC=flag.isChecked(Flag.Parameter.DC);
        short[][]buf;//=new short[SIZEOFBLOCK][SIZEOFBLOCK];

        for (int i = 0; i < duWidth; i++) {
            for (int j = 0; j < duHeight; j++) {//j=3 erro

                buf= OPCMultiThread.getDataOrigin(dopc[i][j], flag);

                for (int x = 0; x < DCTMultiThread.SIZEOFBLOCK; x++) {
                    for (int y = 0; y < DCTMultiThread.SIZEOFBLOCK; y++) {

                        int curX=i * DCTMultiThread.SIZEOFBLOCK + x;
                        int curY=j * DCTMultiThread.SIZEOFBLOCK + y;
                        if (curX< Width && curY < Height)
                            res[curX][curY] = buf[x][y];
                        // DU[i][j].setValue(val,x,y);

                    }
                }


            }
        }
        return res;
    }

    private DataOpc[][] directOpcGlobalBase(int n, int m, short[][]dataOrigin){
        DataOpc[][]dopc=createDataOpc(dataOrigin);

        findAllBase(dataOrigin,dopc);

        setMaxBaseForAll(n, m,dopc);

        directOPCwithGlobalBase(dataOrigin, dopc);

        return dopc;
    }
    private DataOpc[][] createDataOpc(short[][]dataOrigin){
        Size size=sizeOpcCalculate(dataOrigin.length, dataOrigin[0].length);
        int duWidth= size.getWidth();
        int duHeight= size.getHeight();

        DataOpc[][]dopc=new DataOpc[duWidth][duHeight];
        return dopc;
    }
    private void findAllBase(short[][]dataOrigin,DataOpc[][]dopc){
        short[][]buf=new short[SIZEOFBLOCK][SIZEOFBLOCK];

        int duWidth=dopc.length;
        int duHeight=dopc[0].length;
        int Width=dataOrigin.length;
        int Height=dataOrigin[0].length;


        for (int i = 0; i < duWidth; i++) {
            for (int j = 0; j < duHeight; j++) {

                for (int x = 0; x < DCTMultiThread.SIZEOFBLOCK; x++) {
                    for (int y = 0; y < DCTMultiThread.SIZEOFBLOCK; y++) {
                        short value = 0;
                        int curX=i * DCTMultiThread.SIZEOFBLOCK + x;
                        int curY=j * DCTMultiThread.SIZEOFBLOCK + y;
                        if (curX< Width && curY < Height)
                            value = dataOrigin[curX][curY];
                        buf[x][y] = value;
                        // DU[i][j].setValue(val,x,y);

//                        if(x!=0||y!=0)
//                            assert value<0xff:"value["+curX+"]["+curY+"]="+value;
                    }
                }
                dopc[i][j]= OPCMultiThread.findBase(buf, flag);

            }
        }
    }
    private void setMaxBaseForAll(int n,int m,DataOpc[][]dopc){
        int duWidth=dopc.length;
        int duHeight=dopc[0].length;
        int i = 0;
        int j = 0;
        int IndexI=i,IndexJ=j;

        for (; i < duWidth; IndexI+=n,i=IndexI,IndexJ=0,j=IndexJ) {
            for (; j < duHeight; i=IndexI,IndexJ+=m,j=IndexJ) {

                short [] maxBase=new short[SIZEOFBLOCK];
                for(int a=0;a<n&&i<duWidth;a++,i++){
                    j=IndexJ;
                    for(int b=0;b<m&&j<duHeight;b++,j++){
                        maxBase=findMaxInArry(maxBase,dopc[i][j].getBase());
                    }
                }

                i=IndexI;
                for(int a=0;a<n&&i<duWidth;a++,i++){
                    j=IndexJ;
                    for(int b=0;b<m&&j<duHeight;b++,j++){
                        dopc[i][j].setBase(maxBase);
                    }
                }
            }
        }
    }
    private short[] findMaxInArry(short[] a,short[] b){
        assert (a.length==b.length);
        for(int i=0;i<a.length;i++){
            a[i]=(a[i]>b[i])?a[i]:b[i];
        }
        return a;
    }
    private void directOPCwithGlobalBase(short[][]dataOrigin,DataOpc[][]dopc){
        short[][]buf=new short[SIZEOFBLOCK][SIZEOFBLOCK];
        int duWidth=dopc.length;
        int duHeight=dopc[0].length;
        int Width=dataOrigin.length;
        int Height=dataOrigin[0].length;
        for (int i = 0; i < duWidth; i++) {
            for (int j = 0; j < duHeight; j++) {

                for (int x = 0; x < DCTMultiThread.SIZEOFBLOCK; x++) {
                    for (int y = 0; y < DCTMultiThread.SIZEOFBLOCK; y++) {
                        short value = 0;
                        int curX=i * DCTMultiThread.SIZEOFBLOCK + x;
                        int curY=j * DCTMultiThread.SIZEOFBLOCK + y;
                        if (curX< Width && curY < Height)
                            value = dataOrigin[curX][curY];
                        buf[x][y] = value;
                        // DU[i][j].setValue(val,x,y);
                    }
                }
                dopc[i][j]= OPCMultiThread.directOPCwithFindedBase(buf,dopc[i][j], flag);

            }
        }
    }

    private Size sizeOpcCalculate(int Width, int Height) {
        int widthOPC = Width / DCTMultiThread.SIZEOFBLOCK;
        int heightOPC = Height / DCTMultiThread.SIZEOFBLOCK;
        if (Width % DCTMultiThread.SIZEOFBLOCK != 0)
            widthOPC++;
        if (Height % DCTMultiThread.SIZEOFBLOCK != 0)
            heightOPC++;
        //   createMatrix();
        return new Size(widthOPC,heightOPC);
    }

    public short[][] getDataOrigin() {
        if(state==State.Opc&&!isReady) {
            dataOrigin = reverceOPC(dataOpcs);
            isReady=true;
        }

        return dataOrigin;
    }

    public DataOpc[][] getDataOpcs() {
        if(state==State.Origin&&!isReady) {
            dataOpcs = directOPC(dataOrigin);
            isReady=true;
        }

        return dataOpcs;
    }

    /**
     * calculate(if need) DataOpcs with global base for (nxm)
     * @param n - vertical size of same base
     * @param m - horizonlat size of same base
     * @return matrix of DataOpc with same base 
     */
    public DataOpc[][] getDataOpcs(int n,int m){
        if(state==State.Origin&&!isReady) {
            dataOpcs = directOpcGlobalBase(n,m,dataOrigin);
            isReady=true;
        }

        return dataOpcs;
    }
}

File: ImageCompressionLib/src/ImageCompressionLib/utils/objects/OpcConvertor.kt
package ImageCompressionLib.utils.objects

import ImageCompressionLib.containers.*
import ImageCompressionLib.containers.matrix.DataOpcMatrix
import ImageCompressionLib.containers.matrix.Matrix
import ImageCompressionLib.containers.matrix.ShortMatrix
import ImageCompressionLib.containers.type.ByteVector
import ImageCompressionLib.containers.type.DataOpc
import ImageCompressionLib.containers.type.Flag
import ImageCompressionLib.containers.type.Size
import ImageCompressionLib.utils.functions.opc.OpcProcess
import ImageCompressionLib.utils.functions.opc.OpcUtils

class OpcConvertor {
    enum class State {
        Opc, Origin
    }
    //TODO replace forEach with for()
    //TODO do all process in one loop

    private val shortMatrix: Matrix<Short>
    private val dataOpcMatrix: DataOpcMatrix
    private val parameters:Parameters
    var isReady = false
        private set
    var state: State
        private set

    private lateinit var splitedShortMatrix: Matrix<Matrix<Short>>
//    private lateinit var splitedDataOpcMatrix:Matrix<DataOpcMatrix>//TODO make local


    constructor(dataOrigin: Array<Array<Short>>, parameters: Parameters) {
        this.shortMatrix = ShortMatrix.valueOf(dataOrigin)
        this.parameters=parameters
        state = State.Origin
        val size=calculataDataOpcMatrixSize(Size(dataOrigin.size,dataOrigin[0].size),parameters.unitSize)
        dataOpcMatrix= DataOpcMatrix(size.width, size.height, parameters.unitSize)
    }
    constructor(dataOrigin: ShortMatrix, parameters: Parameters) {
        this.shortMatrix = dataOrigin
        this.parameters=parameters
        state = State.Origin
        val size=calculataDataOpcMatrixSize(dataOrigin.size,parameters.unitSize)
        dataOpcMatrix= DataOpcMatrix(size.width, size.height, parameters.unitSize)
    }


    constructor(dataOpcMatrix: DataOpcMatrix, parameters: Parameters){
        this.dataOpcMatrix= (dataOpcMatrix)
        this.parameters=parameters
        state = State.Opc
        val size=calculateShortMatrixSize(dataOpcMatrix.size,parameters.unitSize)
        shortMatrix= ShortMatrix(size.width, size.height)
    }
    private fun createSplitedMatrix(){
        if(!parameters.flag.isChecked(Flag.Parameter.Enlargement))
            splitedShortMatrix=shortMatrix.split(parameters.unitSize.width,parameters.unitSize.height)
        else
            splitedShortMatrix=shortMatrix.splitWithZeroIterator(parameters.unitSize.width,parameters.unitSize.height,0)
    }
    private fun calculataDataOpcMatrixSize(imageSize: Size, unitSize:Size): Size {
        var w= imageSize.width/unitSize.width
        var h= imageSize.height/unitSize.height
        if(imageSize.width%unitSize.width!=0)w++
        if(imageSize.height%unitSize.height!=0)h++
        return Size(w, h)
    }
    private fun calculateShortMatrixSize(dataOpcMatrixSize:Size,unitSze:Size): Size {
        return Size(dataOpcMatrixSize.width*unitSze.width,dataOpcMatrixSize.height*unitSze.height)
    }
    private fun beforDirectOpc(){
        dataOpcMatrix.forEach() { i, j, value ->
            OpcProcess.preDirectOpcProcess(parameters, splitedShortMatrix[i, j], value)
            return@forEach null
        }
    }
    private fun afterReverceOpc(){
        dataOpcMatrix.forEach() { i, j, value ->
            OpcProcess.afterReverceOpcProcess(parameters,value, splitedShortMatrix[i, j])
            return@forEach null
        }
    }
    private fun setGlobalBase(){
//        if(!parameters.flag.isChecked(Flag.Parameter.GlobalBase))
//            return
        val splitedDataOpcMatrix=dataOpcMatrix.split(parameters.sameBaseSize.width,parameters.sameBaseSize.height)
        splitedDataOpcMatrix.forEach(){i, j, value ->
            OpcUtils.setSameBaseIn(value)
            return@forEach null
        }
    }
    private fun directOpc(){
        dataOpcMatrix.forEach(){i, j, value ->
            OpcProcess.directOPC(parameters, splitedShortMatrix[i,j],value)
            return@forEach null
        }
    }
    private fun reverceOPC(){
        dataOpcMatrix.forEach(){i, j, value ->
            OpcProcess.reverseOPC(parameters,value, splitedShortMatrix[i,j])
            return@forEach null
        }
    }
    private fun directOpcWithMessageAt(encParameters: EncryptParameters, message: ByteVector){
        if(encParameters.steganography==null)
            throw Exception("steganography==null")
        val position= encParameters.steganography!!.stegoPosition
        val stegoGeter= encParameters.steganography!!.stegoBlockKeygenFactory.invoke()
        dataOpcMatrix.forEach(){i, j, value ->
            if(message.hasNextBit()&&stegoGeter.isUseNextBlock())
                OpcProcess.directOpcWithMessageAt(parameters, splitedShortMatrix[i,j] ,value,message.getNextBoolean(),position)
            else
                OpcProcess.directOpcWithMessageAt(parameters, splitedShortMatrix[i,j] ,value,false,position)
            return@forEach null
        }
    }
    private fun reverceOPCWithMessageAt(encParameters: EncryptParameters): ByteVector {
        val res= ByteVector()
        if(encParameters.steganography==null)
            throw Exception("steganography==null")
        val position= encParameters.steganography!!.stegoPosition
        val stegoGeter= encParameters.steganography!!.stegoBlockKeygenFactory.invoke()
        dataOpcMatrix.forEach{i, j, value ->
            val tmp=OpcProcess.reverseOpcWithMessageAt(parameters,value, splitedShortMatrix[i,j] ,position)
            if(stegoGeter.isUseNextBlock())res.append(tmp)
            return@forEach null
        }
        encParameters.message=res
        return res
    }
    private fun directProcess(encParameters: EncryptParameters?, message: ByteVector?){
        createSplitedMatrix()

        beforDirectOpc()

        if(parameters.flag.isChecked(Flag.Parameter.GlobalBase))
            setGlobalBase()

        if(encParameters?.steganography!=null&&message!=null)
            directOpcWithMessageAt(encParameters,message)
        else
            directOpc()
    }
    private fun reverceProcess(encParameters: EncryptParameters?): ByteVector?{
        createSplitedMatrix()

        var res: ByteVector?=null
        if(encParameters?.steganography!=null)
            res = reverceOPCWithMessageAt(encParameters)
        else
            reverceOPC()

        afterReverceOpc()

        return res
    }


    fun getDataOrigin(encParameters: EncryptParameters?=null): Pair<Matrix<Short>,ByteVector?> {
        var m:ByteVector?=null
        if (state == State.Opc && !isReady) {
            m=reverceProcess(encParameters)
            isReady = true
        }

        return Pair(shortMatrix,m)
    }

    /**
     * calculate(if need) DataOpcs with global base for (nxm)
     * @param n - vertical size of same base
     * @param m - horizonlat size of same base
     * @return matrix of DataOpcOld with same base
     */
    fun getDataOpcs(): Matrix<DataOpc> {
        if (state == State.Origin && !isReady) {
            directProcess(null,null)
            isReady = true
        }
        return dataOpcMatrix
    }
    fun getDataOpcs(encParameters: EncryptParameters?,message:ByteVector?): Matrix<DataOpc> {
        if (state == State.Origin && !isReady) {
            directProcess(encParameters,message)
            isReady = true
        }
        return dataOpcMatrix
    }
}